dbms/src/Functions/FunctionsLogical.h

```diff
@@ -377,154 +504,421 @@ class FunctionAnyArityLogical : public IFunction
+                }
+            }
+            else
             {
-                vec_input_has_null.assign(rows, (UInt8)0);
-                vec_res_not_null.assign(rows, (UInt8)0);
+                // case 3: 2 nullable
+                NullableAssociativeOperationImpl<Impl, 2>::execute(
+                    nullable_uint8_columns,
+                    null_maps,
+                    vec_res,
+                    vec_res_is_null);
             }
         }
+        if (result_is_nullable)
+        {
+            block.getByPosition(result).column = ColumnNullable::create(std::move(col_res), std::move(col_res_is_null));
+        }
+        else
+        {
+            block.getByPosition(result).column = std::move(col_res);
+        }
+    }
 
-        /// Convert all columns to UInt8
-        UInt8ColumnPtrs uint8_in;
-        Columns converted_columns;
-
-        for (size_t index = 0; index < in.size(); index++)
+    void handleMultipleInputColumns(
+        Block & block,
+        size_t result,
+        size_t rows,
+        bool has_const,
+        UInt8 const_val,
+        UInt8 const_val_is_null,
+        UInt8ColumnPtrs & not_null_uint8_columns,
+        UInt8ColumnPtrs & nullable_uint8_columns,
+        std::vector<const UInt8Container *> & null_maps,
+        bool result_is_nullable) const
+    {
+        // logical op with more than 2 input columns
+        auto col_res = ColumnUInt8::create(rows, static_cast<UInt8>(0));
+        UInt8Container & vec_res = col_res->getData();
+        auto col_res_is_null = ColumnUInt8::create();
+        UInt8Container & vec_res_is_null = col_res_is_null->getData();
+        if (result_is_nullable)
+            vec_res_is_null.assign(rows, static_cast<UInt8>(0));
+
+        bool has_not_null_column;
+        // first handle not-null column
+        if (!not_null_uint8_columns.empty())
         {
-            const IColumn * column = in[index];
-            bool is_const_column [[maybe_unused]] = (Int32)index == const_column_index;
-            if (auto uint8_column = checkAndGetColumn<ColumnUInt8>(column))
+            has_not_null_column = true;
+            if (not_null_uint8_columns.size() == 1)
             {
-                uint8_in.push_back(uint8_column);
-                const auto & data = uint8_column->getData();
-                if constexpr (special_impl_for_nulls)
+                vec_res.assign(not_null_uint8_columns[0]->getData());
+            }
+            else
+            {
+                while (not_null_uint8_columns.size() > 1)
                 {
-                    if (!is_const_column)
-                    {
-                        size_t n = uint8_column->size();
-                        for (size_t i = 0; i < n; i++)
-                            vec_res_not_null[i] |= Impl::resNotNull(data[i], false);
-                    }
+                    AssociativeOperationImpl<Impl, 6>::execute(not_null_uint8_columns, vec_res);
```

Comment:
Keep previous comments to explain why uses 6 here?

Reply:
Done